### PR TITLE
ATO-576: Create processing identity lambda

### PR DESCRIPTION
**Pull request by gn-GDS**
## Checklist

- [ ] Impact on orch and auth mutual dependencies has been checked.

- [ ] A UCD review has been performed.

### CODE REVIEW

File: template.yaml [hunk 1]
```diff
@@ -73,6 +80,13 @@ Mappings:
       userProfileTableKeyArn: arn:aws:kms:eu-west-2:761723964695:key/12f40ae0-84a0-4840-a497-129366eef354
       txmaEventEncryptionKeyArn: arn:aws:kms:eu-west-2:767397776536:key/4d851d1e-acd1-4c73-a754-617ffd380b3d
       orchToAuthTokenSigningKeyAlias: arn:aws:kms:eu-west-2:761723964695:key/1800ecc2-e04d-4cf5-9b4e-72eafa0c71f6
+      accountModifiersTableKeyArn: arn:aws:kms:eu-west-2:761723964695:key/4932c4d3-bf86-4f39-a32c-b82faf9a2574
+      userCredentialsTableKeyArn: arn:aws:kms:eu-west-2:761723964695:key/13c5043c-3c9e-4370-bff6-b70c2d8bc609
+      identityCredentialsTableKeyArn: arn:aws:kms:eu-west-2:761723964695:key/b1bc6357-0578-45cc-a192-50e609094b4e
+      accountInterventionServiceAbortOnError: false
+      accountInterventionServiceActionEnabled: false
+      accountInterventionServiceCallEnabled: false
```
Comment (CarlyG55) [inline template.yaml:88]: pretty sure some of these values are wrong - do you mind double checking them for each environment (or you can check my pr for ipv callback and check that I've done it right 😆 )

File: template.yaml [hunk 2]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+          ACCOUNT_INTERVENTION_SERVICE_ABORT_ON_ERROR: !FindInMap [ EnvironmentConfiguration, !Ref Environment, accountInterventionServiceAbortOnError ]
+          ACCOUNT_INTERVENTION_SERVICE_CALL_ENABLED:  !FindInMap [ EnvironmentConfiguration, !Ref Environment, accountInterventionServiceCallEnabled ]
+          ACCOUNT_INTERVENTION_SERVICE_CALL_TIMEOUT: !FindInMap [ EnvironmentConfiguration, !Ref Environment, accountInterventionServiceCallTimeout]
+          ACCOUNT_INTERVENTION_SERVICE_ACTION_ENABLED: !FindInMap [ EnvironmentConfiguration, !Ref Environment, accountInterventionServiceActionEnabled ]
+          TXMA_AUDIT_QUEUE_URL:
+            Fn::ImportValue: !Sub orchestration-${Environment}-txma-QueueURL
+          OIDC_API_BASE_URL: !Sub
+            - https://oidc.${ServiceDomain}/
+            - ServiceDomain: !FindInMap [ EnvironmentConfiguration, !Ref Environment, serviceDomain ]
+
```
Comment (CarlyG55) [inline template.yaml:1384]: will need the dynamo arn prefix

Comment (CarlyG55) [reply template.yaml:1384]: also the redis key and frontend base url

File: template.yaml [hunk 3]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+            Effect: Allow
+            Action:
+              - kms:Encrypt
+              - kms:Decrypt
+              - kms:ReEncrypt*
+              - kms:GenerateDataKey*
+              - kms:CreateGrant
+              - kms:DescribeKey
+            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, accountModifiersTableKeyArn ]
+
```
Comment (CarlyG55) [inline template.yaml:1539]: I think this is a mistake in the terraform and it's not actually needed

File: template.yaml [hunk 4]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+          - Sid: AllowUserProfileTableReadAndWriteAccess
+            Effect: "Allow"
+            Action:
+              - dynamodb:DescribeTable
+              - dynamodb:Get*
+              - dynamodb:Query
+              - dynamodb:Scan
+              - dynamodb:UpdateItem
+              - dynamodb:PutItem
```
Comment (CarlyG55) [inline template.yaml:1508]: Don't need user profile write access

File: template.yaml [hunk 5]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+          - Sid: AllowIdentityCredentialsTableAccess
+            Effect: "Allow"
+            Action:
+              - dynamodb:BatchGetItem
+              - dynamodb:DescribeTable
+              - dynamodb:Get*
+              - dynamodb:Query
+              - dynamodb:Scan
+              - dynamodb:UpdateItem
+              - dynamodb:PutItem
```
Comment (CarlyG55) [inline template.yaml:1459]: Don't need write access

File: template.yaml [hunk 6]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+          - Sid: AllowUserCredentialsTableKms
+            Effect: Allow
+            Action:
+              - kms:Encrypt
+              - kms:Decrypt
+              - kms:ReEncrypt*
+              - kms:GenerateDataKey*
+              - kms:CreateGrant
+              - kms:DescribeKey
+            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, userCredentialsTableKeyArn ]
```
Comment (CarlyG55) [inline template.yaml:1448]: Don't need these, the terraform roles were a bit too lax

File: template.yaml [hunk 7]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+        CheckovRulesToSkip: CKV_AWS_116.CKV_AWS_117.CKV_AWS_173
+
+  ProcessingIdentityFunctionCrossAccountPermission:
+    Type: AWS::Lambda::Permission
+    Properties:
+      FunctionName: !Ref ProcessingIdentityFunction.Alias
+      Action: lambda:InvokeFunction
+      Principal: apigateway.amazonaws.com
+      SourceArn: !Sub
+        - 'arn:aws:execute-api:eu-west-2:${AccountId}:${ApiId}/*/GET/.well-known/jwks.json'
```
Comment (CarlyG55) [inline template.yaml:1566]: this needs updating for ideneity processing

File: template.yaml [hunk 8]
```diff
@@ -1317,6 +1351,298 @@ Resources:
...
+        - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+          ApiId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authApiId ]
+
+  ProcessingIdentityFunctionLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Sub /aws/lambda/${Environment}-jwks-lambda
```
Comment (CarlyG55) [inline template.yaml:1573]: needs updating

Comment (CarlyG55) [reply template.yaml:1573]: I think there's still quite a few jwks things copied over if you could double check

File: ci/terraform/oidc/api-gateway.tf [hunk 9]
```diff
@@ -1020,4 +1022,39 @@ resource "aws_api_gateway_integration" "orch_authorisation_integration" {
...
+  rest_api_id = aws_api_gateway_rest_api.di_authentication_api.id
+  resource_id = aws_api_gateway_resource.orch_processing_identity_resource[0].id
+  http_method = aws_api_gateway_method.orch_processing_identity_method[0].http_method
+  depends_on = [
+    aws_api_gateway_resource.orch_processing_identity_resource
+  ]
+  type                    = "AWS_PROXY"
+  integration_http_method = "POST"
+  uri                     = "arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:${var.orch_account_id}:function:${var.orch_processing_identity_name}:latest/invocations"
+}
```
Comment (isaac-GDS) [inline ci/terraform/oidc/api-gateway.tf:1060]: Is this correct here ? This lambda is exposed on the frontend api ? If so will need to be moved and rest_api_id corrected to reference frontend.

File: template.yaml [hunk 10]
```diff
@@ -136,7 +152,10 @@ Mappings:
       userProfileTableKeyArn: arn:aws:kms:eu-west-2:172348255554:key/365c4de5-6a6d-43db-8569-eca00e889177
       txmaEventEncryptionKeyArn: arn:aws:kms:eu-west-2:533266965190:key/4bc0dac2-f40f-4c3e-9f9e-30549dd8769d
       orchToAuthTokenSigningKeyAlias: arn:aws:kms:eu-west-2:172348255554:key/cb986cd8-8ac7-43f1-8a74-15d1b77509d9
-
+      identityCredentialsTableKeyArn: arn:aws:kms:eu-west-2:172348255554:key/36023c7d-ec27-45c8-9edd-0761974dd0d3
+      accountInterventionServiceAbortOnError: false
+      accountInterventionServiceActionEnabled: true
+      accountInterventionServiceCallEnabled: true
 Globals:
   Function:
```
Comment (hannah-leong) [inline template.yaml:160]:
```suggestion
      identityCredentialsTableKeyArn: arn:aws:kms:eu-west-2:172348255554:key/36023c7d-ec27-45c8-9edd-0761974dd0d3
      accountInterventionServiceAbortOnError: false
      accountInterventionServiceActionEnabled: true
      accountInterventionServiceCallEnabled: true
      
Globals:
  Function:
```

File: template.yaml [hunk 11]
```diff
@@ -1317,6 +1336,267 @@ Resources:
...
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
+          - Sid: AllowDecryptOfParameters
+            Effect: "Allow"
+            Action: "kms:Decrypt"
+            Resource: !GetAtt MainKmsKey.Arn
+
+          - Sid: AllowClientRegistryTableReadAccess
+            Effect: "Allow"
+            Action:
+              - dynamodb:BatchGetItem
```
Comment (hannah-leong) [inline template.yaml:1398]: Tiny nitpick but maybe remove quotation marks for consistency?

**Comment by ethanmills** [comment]
@gn-GDS shall we close this one for now and revisit it when we're ready to move this?

**Comment by ethanmills** [comment]
Closing until we're ready to move this

**Pull request closed without merging**